File: backend/app/core/file_restrictions.py
from pathlib import Path

from fastapi import HTTPException, status


RESTRICTED_FILE_EXTENSIONS = {
    ".exe",
    ".msi",
    ".bat",
    ".cmd",
    ".com",
    ".scr",
    ".js",
    ".vbs",
    ".ps1",
    ".sh",
    ".py",
    ".php",
    ".html",
    ".htm",
    ".jsp",
    ".asp",
    ".aspx",
}


def get_file_extension(filename: str | None) -> str:
    return Path(str(filename or "").strip()).suffix.lower()


def ensure_allowed_filename(filename: str | None) -> str:
    normalized_name = str(filename or "").strip()
    ext = get_file_extension(normalized_name)
    if ext in RESTRICTED_FILE_EXTENSIONS:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=f"Files with the '{ext}' extension are not allowed.",
        )
    return normalized_name

File: frontend/src/lib/fileRestrictions.js
const RESTRICTED_FILE_EXTENSIONS = new Set([
  '.exe',
  '.msi',
  '.bat',
  '.cmd',
  '.com',
  '.scr',
  '.js',
  '.vbs',
  '.ps1',
  '.sh',
  '.py',
  '.php',
  '.html',
  '.htm',
  '.jsp',
  '.asp',
  '.aspx',
])

export const getFileExtension = (name = '') => {
  const normalized = String(name || '').trim()
  const index = normalized.lastIndexOf('.')
  return index >= 0 ? normalized.slice(index).toLowerCase() : ''
}

export const isRestrictedFileType = (fileOrName) => {
  const name = typeof fileOrName === 'string' ? fileOrName : fileOrName?.name
  return RESTRICTED_FILE_EXTENSIONS.has(getFileExtension(name))
}

export const getRestrictedFileTypeMessage = (fileOrName) => {
  const name = typeof fileOrName === 'string' ? fileOrName : fileOrName?.name
  const ext = getFileExtension(name)
  return ext
    ? `Files with the '${ext}' extension are not allowed.`
    : 'This file type is not allowed.'
}
